### PR TITLE
UGENE-7508. Align selected sequence to Alignment with MUSCLE should not activate Undo when nothing is changed

### PR DESCRIPTION
**Pull request by yalgaer**

### CODE REVIEW

File: src/corelibs/U2Core/src/datatype/msa/MultipleAlignment.cpp [hunk 1]
```diff
@@ -55,10 +55,11 @@ const MultipleAlignmentData *MultipleAlignment::operator->() const {
...
 
-MultipleAlignmentData::MultipleAlignmentData(const QString &name, const DNAAlphabet *alphabet, const QList<MultipleAlignmentRow> &rows)
-    : alphabet(alphabet),
-      rows(rows),
-      length(0) {
+MultipleAlignmentData::MultipleAlignmentData(const MultipleAlignmentDataType &_type,
+                                             const QString &name,
+                                             const DNAAlphabet *_alphabet,
+                                             const QList<MultipleAlignmentRow> &_rows)
+    : type(_type), alphabet(_alphabet), rows(_rows), length(0) {
```
Comment (DmitriiSukhomlinov) [inline src/corelibs/U2Core/src/datatype/msa/MultipleAlignment.cpp]: You can set a default value to `length` in the .h file and do not set it here

File: src/corelibs/U2Core/src/datatype/msa/MultipleAlignment.h [hunk 2]
```diff
@@ -90,12 +90,13 @@ class U2CORE_EXPORT MultipleAlignmentData {
...
      */
-    MultipleAlignmentData(const QString &name = QString(),
+    MultipleAlignmentData(const MultipleAlignmentDataType &type,
+                          const QString &name = QString(),
                           const DNAAlphabet *alphabet = nullptr,
                           const QList<MultipleAlignmentRow> &rows = QList<MultipleAlignmentRow>());
 
 public:
-    virtual ~MultipleAlignmentData();
+    virtual ~MultipleAlignmentData() = default;
```
Comment (DmitriiSukhomlinov) [inline src/corelibs/U2Core/src/datatype/msa/MultipleAlignment.h:99]: You can remove the declaration at all, I suppose

File: src/corelibs/U2Core/src/datatype/msa/MultipleChromatogramAlignmentRow.cpp [hunk 3]
```diff
@@ -400,56 +400,31 @@ QPair<DNAChromatogram::ChromatogramTraceAndValue, DNAChromatogram::ChromatogramT
     return {peaks[0], peaks[1]};
 }
 
-bool MultipleChromatogramAlignmentRowData::isRowContentEqual(const MultipleChromatogramAlignmentRow &row) const {
-    return isRowContentEqual(*row);
+bool MultipleChromatogramAlignmentRowData::isEqual(const MultipleAlignmentRowData &other) const {
+    CHECK(other.type == MultipleAlignmentDataType::MCA, false);
+    auto mcaRow = dynamic_cast<const MultipleChromatogramAlignmentRowData *>(&other);
+    SAFE_POINT(mcaRow != nullptr, "Not an MCA row!", false);
```
Comment (DmitriiSukhomlinov) [inline src/corelibs/U2Core/src/datatype/msa/MultipleChromatogramAlignmentRow.cpp:406]: Use `L10N::nullPointer` instead

File: src/corelibs/U2Core/src/datatype/msa/MultipleChromatogramAlignmentRow.cpp [hunk 4]
```diff
@@ -400,56 +400,31 @@ QPair<DNAChromatogram::ChromatogramTraceAndValue, DNAChromatogram::ChromatogramT
...
+}
 
-            if (firstRowGaps == secondRowGaps) {
-                return true;
-            }
-        }
+bool MultipleChromatogramAlignmentRowData::isEqualCore(const MultipleAlignmentRowData &other) const {
+    if (other.type == MultipleAlignmentDataType::MCA) {
+        auto mcaRow = dynamic_cast<const MultipleChromatogramAlignmentRowData *>(&other);
+        SAFE_POINT(mcaRow != nullptr, "Not an MCA row!", false);
```
Comment (DmitriiSukhomlinov) [inline src/corelibs/U2Core/src/datatype/msa/MultipleChromatogramAlignmentRow.cpp:419]: The same

**Review by DmitriiSukhomlinov — CHANGES_REQUESTED** [review]
Some minor comments

File: src/corelibs/U2Core/src/datatype/msa/MultipleAlignment.cpp (hunk 1 above)
Comment (yalgaer) [reply src/corelibs/U2Core/src/datatype/msa/MultipleAlignment.cpp]: Fixed

File: src/corelibs/U2Core/src/datatype/msa/MultipleChromatogramAlignmentRow.cpp (hunk 3 above)
Comment (yalgaer) [reply src/corelibs/U2Core/src/datatype/msa/MultipleChromatogramAlignmentRow.cpp:406]: This is not a null-ptr error, this is an invalid type error.

File: src/corelibs/U2Core/src/datatype/msa/MultipleChromatogramAlignmentRow.cpp (hunk 4 above)
Comment (yalgaer) [reply src/corelibs/U2Core/src/datatype/msa/MultipleChromatogramAlignmentRow.cpp:419]: Same here: this is not a null-ptr error, this is an invalid type error.

File: src/corelibs/U2Core/src/datatype/msa/MultipleAlignment.h (hunk 2 above)
Comment (yalgaer) [reply src/corelibs/U2Core/src/datatype/msa/MultipleAlignment.h:99]: Why? I want to declare that the destructor in this hierarchy is virtual
https://newbedev.com/virtual-default-destructors-in-c

**Pull request merged**